proposals/0181-privileged-syscalls.md

```diff
@@ -0,0 +1,101 @@
+---
+simd: '0181'
+title: Privileged Syscalls
+authors:
+  - Joe Caulfield | Anza
+category: Standard
+type: Core
+status: Draft
+created: 2024-10-02
+feature: (fill in with feature tracking issues once accepted)
+supersedes: (optional - fill this in if the SIMD supersedes a previous SIMD)
+superseded-by: (optional - fill this in if the SIMD is superseded by a subsequent
+ SIMD)
+extends: (optional - fill this in if the SIMD extends the design of a previous
+ SIMD)
+---
+
+## Summary
+
+Support for privileged syscalls, to perform special-cased runtime activities,
+for enshrined program IDs only.
+
+## Motivation
+
+As we continue the effort to migrate all builtin programs to on-chain BPF
+programs (Core BPF), the ability for some of these programs to perform
+privileged operations becomes immediately relevant.
+
+Some examples of such operations are:
+
+- Allocating very large accounts (System program)
+- Marking accounts as `executable` (loaders)
```

Comment:
This requirement is already being phased out through #162 and #167.

Reply:
That may be, but there are other use cases for these class of syscalls. Some potential examples:
- ELF verification
- Updating the program JIT cache
- Reduced/ignored CUs for vote processing

I can add them to the proposal once I firm up the list of known privileged operations.

Reply:
Agreed that we probably will need a few exceptions to raise limits in the program runtime for core programs. But I don't think that these should come as syscalls. In general we try to avoid adding syscalls and built-ins.

Reply:
The syscall approach I feel like gives us a more standardized interface and specification for these kinds of operations, rather than just baking them into the program runtime's implementation details.

For example, such a specification (privileged syscalls) makes reimplementation with other clients/VMs easier I would think.

What kind of alternatives did you have in mind?